### ISSUE TITLE
[IMPROVEMENT] Reduce backup&restore costs by replacing backup JSON formatted metafile with SQLite database

### ISSUE THREAD

**Issue by derekbit** [longhorn/longhorn#5210]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

The backup information is stored as a JSON metafile, which contains the name of the backup, the snapshot name, blocks, etc. During backup, the block mappings for the volume are temporarily stored in memory and has uncessary memory consumption. Additionally, the JSON formatted metafile is not easy for reading and debugging. The purpose of this task is to replace the JSON-formatted metafile with a SQLite database.

Additionally, there are some advantages using a SQLite database
1. Avoid unnecessary remote checksum lookups. Check the local database to determine whether the block already exists.
2. Restoring using a SQLite database can reduce the number of S3 GET operations. We can easily SELECT the blocks with same checksum and then restore multiple blocks containing the same data by a single downloaded block.

The downside of SQLite is the temporary comsumption of local disk storage.

**Comment by innobead** [comment]
This should be related to https://github.com/longhorn/longhorn/issues/5000 together.

The goal should be to ensure metadata integrity because file-based metadata seems error-prone by user intervene. WDYT?

**Comment by derekbit** [comment]
> This should be related to #5000 together.
> 
> The goal should be to ensure metadata integrity because file-based metadata seems error-prone by user intervention. The JSON meta is still good for debugging and human reading, so this is not our major concern. WDYT?

Hmm, For backup attributes, yes. But for `blocks`, I think no.

The blocks field is like `{"offset": "2551522", "blockChecksum", "abc....de"}, {"offset": "12254331", "blockChecksum", "abc....de"}, {"offset": "32251511", "blockChecksum", "abc....de"}, {"offset": "0", "blockChecksum", "abc....de"}....{"offset": "11255422", "blockChecksum", "abc....de"}"`

**Comment by innobead** [comment]
> > The JSON meta is still good for debugging and human reading, so this is not our major concern. WDYT?
> 
> Hmm, For backup attributes, yes. But for `blocks`, I think no.
> 
> The blocks field is like `{"offset": "2551522", "blockChecksum", "abc....de"}, {"offset": "12254331", "blockChecksum", "abc....de"}, {"offset": "32251511", "blockChecksum", "abc....de"}, {"offset": "0", "blockChecksum", "abc....de"}....{"offset": "11255422", "blockChecksum", "abc....de"}"`.

Yeah, I see. How would you like to improve this if using SQLite? How will it look like?

**Comment by derekbit** [comment]
Idealy, the metadata db will have two tables, one is for  backup attributes, and the other one is for blocks.
The schemas are simple for the two tables.

I thnk the importance of the blocks table can help the issues I memtioned in https://github.com/longhorn/longhorn/issues/5210#issue-1520049923, but need a siginificant refactoring in backupstore...

**Comment by joshimoo** [comment]
Where would the SQLite db be stored and kept in sync between 2 active clusters?
Is it only supposed to act as a cache on the local cluster?
